ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#include <algorithm>
+#include <cinttypes>
+#include <cmath>
+#include <cstdlib>
+#include <memory>
+#include <random>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+#include "ngraph/function.hpp"
+#include "ngraph/ngraph.hpp"
+#include "util/engine/test_engines.hpp"
+#include "util/test_case.hpp"
+#include "util/test_control.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace
+{
+    // Tolerance can be specified in two forms (number of bits in mantissa or float-point
+    // threshold), corresponding to two different comparing methods implemented in TestCase
+    // classes. Following class acts as an encapsulation on these two options without introducing
+    // too many overloadings on the function requiring such argument.
+    struct ArgTolerance
+    {
+        const int tolerance_bits;
+        const float tolerance_fp;
+        bool using_bits;
+        ArgTolerance()
+            : tolerance_bits(-1)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const int tolerance_bits)
+            : tolerance_bits(tolerance_bits)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const float tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+        ArgTolerance(const double tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+    };
+
+    // ArgShape is used implicitly by caller of test_unary() in following ways:
+    //
+    //  {PartialShape{2,Dimension::dynamic(),3}, Shape{2, 4, 3}}
+    //        A full specified form with both partial shape for ngraph function
+    //        generation and a static shape for final input tensor.
+    //
+    //  Shape{2, 4, 3}
+    //        A pure static shape for both ngraph function generation and input tensor
+    //
+    //  {}
+    //        A unspecified shape, test_unary() will derive the actual shape from the size
+    //        of the input vector as 1D shape.
+    struct ArgShape
```

Comment:
I don't understand the concept of this union. PartialShape is a superset of Shape meaning a fully static Shape can be stored in PartialShape instance. You can use `PartialShape::to_shape()` to retrieve it. 
Also I could not find a spot where this class is instantiated with the first constructor. It's super confusing and I don't see the reason why this struct is needed. Can you explain this a bit more?

Reply:
It's not a union, both two shapes are required, one is of type PartialShape used to define the ngraph Parameter, another is of type Shape used to create input tensor and feed into TestCase. In some test cases they are equal, but they are different when testing on dynamic shape, grep "dynamic" under backend test folder you can see them:

https://github.com/openvinotoolkit/openvino/blob/10af23a90c6ca966e578074c9cdaee403c29d4f2/ngraph/test/backend/non_zero.in.cpp#L19

As for why I put them into one class, it's for reducing the number of overloads for test_unary(). for example if we have 10 function prototypes (be overloads or not) all have arguments requiring such pair of shape, now we want to make caller easier by allowing them to :

1. pass both dynamic shape & static shape for a dynamic shape test
2. pass only one Shape for "static shape" test
3. pass a empty {} for "static shape" test and with 1D shape auto derived from the size of input data vector


we would have to triple the total number of function prototypes (30) if using function overloads, by introducing this class, we don't need any overload on those 10 functions, they all accept arg of ArgShape type, which has 3 constructor overloads, compiler will figure out how to convert user's input to ArgShape by calling the correct constructor.